client-sniffer/src/main/java/org/elasticsearch/client/sniff/Sniffer.java

```diff
@@ -0,0 +1,249 @@
+            this.hostsSniffer = hostsSniffer;
+            this.restClient = restClient;
+            this.sniffInterval = sniffInterval;
+            this.sniffAfterFailureDelay = sniffAfterFailureDelay;
+            this.scheduledExecutorService = Executors.newScheduledThreadPool(1);
+            this.scheduledFuture = this.scheduledExecutorService.schedule(this, 0, TimeUnit.MILLISECONDS);
+            this.nextSniffDelay = sniffInterval;
+        }
+
+        @Override
+        public void run() {
+            sniff(null);
+        }
+
+        void sniffOnFailure(HttpHost failedHost) {
+            this.nextSniffDelay = sniffAfterFailureDelay;
+            sniff(failedHost);
+        }
+
+        void sniff(HttpHost excludeHost) {
+            if (running.compareAndSet(false, true)) {
+                try {
+                    List<HttpHost> sniffedNodes = hostsSniffer.sniffHosts();
+                    if (excludeHost != null) {
+                        sniffedNodes.remove(excludeHost);
+                    }
+                    logger.debug("sniffed nodes: " + sniffedNodes);
+                    this.restClient.setHosts(sniffedNodes.toArray(new HttpHost[sniffedNodes.size()]));
+                } catch (Throwable t) {
+                    logger.error("error while sniffing nodes", t);
+                } finally {
+                    try {
+                        //regardless of whether and when the next sniff is scheduled, cancel it and schedule a new one with updated delay
+                        this.scheduledFuture.cancel(false);
+                        logger.debug("scheduling next sniff in " + nextSniffDelay + " ms");
+                        this.scheduledFuture = this.scheduledExecutorService.schedule(this, nextSniffDelay, TimeUnit.MILLISECONDS);
+                    } catch (Throwable t) {
+                        logger.error("error while scheduling next sniffer task", t);
+                    } finally {
+                        this.nextSniffDelay = sniffInterval;
+                        running.set(false);
+                    }
+                }
+            }
+        }
+
+        void shutdown() {
+            scheduledExecutorService.shutdown();
+            try {
+                if (scheduledExecutorService.awaitTermination(1000, TimeUnit.MILLISECONDS)) {
+                    return;
+                }
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+            scheduledExecutorService.shutdownNow();
+        }
+    }
+
+    /**
+     * Returns a new {@link Builder} to help with {@link Sniffer} creation.
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * Sniffer builder. Helps creating a new {@link Sniffer}.
+     */
+    public static final class Builder {
+        public static final int DEFAULT_SNIFF_INTERVAL = 60000 * 5; //5 minutes
```

Comment:
I think you mean `TimeUnit.MINUTES.toMillis(5)`? I also think the fieldname should include that the units are milliseconds.